engine/storage/src/main/java/org/apache/cloudstack/storage/endpoint/DefaultEndPointSelector.java

```diff
@@ -128,9 +132,24 @@ protected EndPoint findEndPointInScope(Scope scope, String sqlBase, Long poolId)
             }
         }
 
+        List<Long> dedicatedHosts = dedicatedResourceDao.listAllHosts();
+
         // TODO: order by rand() is slow if there are lot of hosts
         sbuilder.append(") t where t.value<>'true' or t.value is null");    //Added for exclude cluster's subquery
-        sbuilder.append(" ORDER by rand() limit 1");
+        sbuilder.append(" ORDER by ");
+        if (dedicatedHosts.size() > 0) {
+            Collections.shuffle(dedicatedHosts); // Randomize dedicated hosts as well.
+            sbuilder.append("field(t.id, ");
+            Iterator<Long> hostIt = dedicatedHosts.iterator();
+            while (hostIt.hasNext()) { // put dedicated hosts at the end of the result set
+                sbuilder.append("'" + hostIt.next() + "'");
+                if (hostIt.hasNext()) {
+                    sbuilder.append(",");
+                }
+            }
+            sbuilder.append(")," );
+        }
```

Comment:
can this be a separate method (i.e. `moveDedicatedHostsToBackForExclusion(sbuilder,dedicatedHosts);`) ?

Reply:
pretty please, @Spaceman1984

Reply:
I'll see what I can do @DaanHoogland

Reply:
Done